File: 54. Spiral Matrix/pop.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        if (matrix.empty()) return res;
        int m = matrix.size(), n = matrix[0].size();
        int up = 0, down = m - 1, left = 0, right = n - 1;
        while (true) {
            for (int col = left; col <= right; col++) res.push_back(matrix[up][col]);
            if (++up > down) break; // reconstruct the boundary, if the boundary is not valid, break
            for (int row = up; row <= down; row++) res.push_back(matrix[row][right]);
            if (--right < left) break;
            for (int col = right; col >= left; col--) res.push_back(matrix[down][col]);
            if (--down < up) break;
            for (int row = down; row >= up; row--) res.push_back(matrix[row][left]);
            if (++left > right) break;
        }
        return res;
    }
};

File: 54. Spiral Matrix/pop.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        while matrix:
            res += matrix.pop(0) # pop the first row, from the first element to the last element
            if matrix and matrix[0]:
                for row in matrix:
                    res.append(row.pop()) # pop the last element of each row, from the first row to the last row
            if matrix:
                res += matrix.pop()[::-1] # pop the last row, from the last element to the first element, and reverse the order
            if matrix and matrix[0]:
                for row in matrix[::-1]:
                    res.append(row.pop(0)) # pop the first element of each row, from the last row to the first row, and reverse the order
        return res